### DISCUSSION TITLE
LoggerSource being able to have its own collection of listeners/sources

### DISCUSSION THREAD

**Discussion by Therzok** [BepInEx/BepInEx#199] [BepInEx 6 proposals]
The current implementation of the Logger forces the usage of a single main Source/Listener collection implementation.

Allowing Sources to have their own listeners to dispatch to, with the default being use the collection the global Logger class uses:

* Allow communication between specific source/listener combinations
* Provide a good default still
* Allow sources to optimize their work based on the registered listeners

This would remove the need to call `LogEvent` for the general use case, thus avoiding allocations of `LogEventArgs` for most cases, unless someone manually subscribes to it.

That would also enable the following scenarios:

1. A temporarily registered logger/listener pair that is toggled when a mod developer needs more logging from a user, without having to ship a new dll - register them as part of `Logger.Sources/Listeners`
2. Create aggregate LogSource implementations with different components of the plugin, which can then be registered/unregistered in bulk.

**Comment by Therzok** [comment]
cc @bbepis @ghorsington @ManlyMarco

**Comment by ghorsington** [comment]
I agree on the general idea here. What would the general API look like in this case?

I'm especially interested in the idea that sources have their own listeners since that would allow to filter outputs better. Right now filtering is kind of difficult since it requires to check the sender instance on listener side (e.g. check DiskLogListener and how it handles filtering out Unity logs). Getting rid of that or simplifying the process would be great.

**Reply by Therzok** [reply]
The way I envision the whole thing is:

```csharp
static class Logger
{
    // not necessarily public
    AggregateLogListener AllListeners;
    ICollection<ILogSource> AllSources;

    void Register(ILogListener listener);
    void Unregister(ILogListener listener);
}

interface ILogSource
{
    AggregateLogListener Listeners; // or APIs to register listeners or listener groups

    string SourceName { get; }

    // I would get rid of this completely replacing it with a method
    // that just dispatches to listeners associated with the source
    event EventHandler<LogEventArgs> LogEvent;

     // alternative:
    void OnLogEvent(string fmt);
}

public interface ILogListener : IDisposable
{
    void LogEvent(string message);
}

// optional: fast-path optimizations that are called to avoid allocations for formatting.
public interface IFormattableLogListener
{
    void LogEvent(string fmt, object arg1);
    void LogEvent(string fmt, object arg1, object arg2);
    void LogEvent(string fmt, object arg1, object arg2, object arg3);
    void LogEvent(string fmt, params object[] args);
}

class AggregateLogListener : ILogListener
{
     // dispatch to collection
}
```

**Comment by Therzok** [comment]
@ghorsington since BepInEx is now 6.0 in the master branch, I'll put up a PR with the proposed changes.

**Reply by ghorsington** [reply]
Sounds good to me!
